pinot-controller/src/main/java/com/linkedin/pinot/controller/api/upload/ZKOperator.java

```diff
@@ -0,0 +1,195 @@
+import com.linkedin.pinot.common.metadata.segment.SegmentZKMetadataCustomMapModifier;
+import com.linkedin.pinot.common.metrics.ControllerMeter;
+import com.linkedin.pinot.common.metrics.ControllerMetrics;
+import com.linkedin.pinot.common.segment.SegmentMetadata;
+import com.linkedin.pinot.common.utils.FileUploadDownloadClient;
+import com.linkedin.pinot.controller.ControllerConf;
+import com.linkedin.pinot.controller.api.resources.ControllerApplicationException;
+import com.linkedin.pinot.controller.api.resources.FileUploadPathProvider;
+import com.linkedin.pinot.controller.helix.core.PinotHelixResourceManager;
+import com.linkedin.pinot.filesystem.PinotFS;
+import com.linkedin.pinot.filesystem.PinotFSFactory;
+import java.io.File;
+import java.net.URI;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.Response;
+import org.apache.helix.ZNRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * The ZKOperator is a util class that is used during segment upload to set relevant metadata fields in zk. It will currently
+ * also perform the data move. In the future when we introduce versioning, we will decouple these two steps.
+ */
+public class ZKOperator {
+  private static final Logger LOGGER = LoggerFactory.getLogger(ZKOperator.class);
+  private PinotHelixResourceManager _pinotHelixResourceManager;
+  private ControllerConf _controllerConf;
+  private ControllerMetrics _controllerMetrics;
+
+  public ZKOperator(PinotHelixResourceManager pinotHelixResourceManager, ControllerConf controllerConf, ControllerMetrics controllerMetrics) {
+    _pinotHelixResourceManager = pinotHelixResourceManager;
+    _controllerConf = controllerConf;
+    _controllerMetrics = controllerMetrics;
+  }
+
+  public void completeSegmentOperations(SegmentMetadata segmentMetadata, URI finalSegmentLocationURI, File currentSegmentLocation, boolean enableParallelPushProtection, HttpHeaders headers,
+      FileUploadPathProvider provider) throws Exception {
+    String rawTableName = segmentMetadata.getTableName();
+    String offlineTableName = TableNameBuilder.OFFLINE.tableNameWithType(rawTableName);
+    String segmentName = segmentMetadata.getName();
+
+    ZNRecord znRecord = _pinotHelixResourceManager.getSegmentMetadataZnRecord(offlineTableName, segmentName);
+
+    // Brand new segment, not refresh, directly add the segment
+    if (znRecord == null) {
+      LOGGER.info("Adding new segment: {}", segmentName);
+      try {
+        moveSegmentToPermanentDirectory(currentSegmentLocation, finalSegmentLocationURI);
+        LOGGER.info("Moved segment {} from temp location {} to {}", segmentName, currentSegmentLocation.getAbsolutePath(), finalSegmentLocationURI.getPath());
+      } catch (Exception e) {
+        LOGGER.error("Could not move segment {} from table {} to permanent directory", segmentName, rawTableName);
+        throw new RuntimeException(e);
+      }
+      if (finalSegmentLocationURI.getScheme().equals("file")) {
```

Comment:
Don't we need to set the download URI irrespective of whether the scheme is "file"?

Reply:
The special case scenario for file is that we will set the downloadUri to the http segment endpoint we download from. Otherwise, we will construct it from controller data dir (with remote storage scheme) and download from there. The downloadUri has been set in the previous calling method. I wanted to put it there because in my mind the controller should come up with that location. However, this downloadUri is reset here because we need to move the final to the constructed final dir first.